Include a task to update nodes with latest available updates

kishen-v opened the pull request:
Introducing a task to update the packages and kernel to the latest available versions.
Currently the task is blocked as it can the time taken to complete depends on the network speeds, but is scoped to be added as a task in the build-cluster creation playbook.

kishen-v committed:
Include a task to update nodes with latest available updates

kishen-v commented on roles/update-pkgs/tasks/main.yml:
```diff
@@ -0,0 +1,27 @@
+- name: Check if reboot required
+  shell: needs-restarting -r
+  register: reboot_check
+  ignore_errors: yes
```
As ansible by default marks the task as fatal, when exit code 1 is observed.

mkumatag commented on roles/update-pkgs/tasks/main.yml:
```diff
@@ -0,0 +1,27 @@
+
+- name: Reboot if necessary
+  reboot:
+    reboot_timeout: 600
```
I see default value is already `600`, so wondering if you want to mention the default value here?

kishen-v replied:
Did away with the defaults..

mkumatag commented on roles/update-pkgs/tasks/main.yml:
```diff
@@ -0,0 +1,27 @@
+- name: Reboot if necessary
+  reboot:
+    reboot_timeout: 600
+    test_command: ping -c 4 {{ ansible_default_ipv4.address }}
```
explanation for this field is `Command to run on the rebooted host and expect success from to determine the machine is ready for further tasks.` and the default command is `whoami`. 

So I'm not sure pining to which ip address here? can we have keep it default or have some command like `uptime` or something else for the confirmation?

mkumatag commented on roles/update-pkgs/tasks/main.yml:
```diff
@@ -0,0 +1,27 @@
+  when: reboot_check.rc == 1 and ansible_distribution in ['CentOS', 'RHEL']
+
+- name: Wait for the system to come back online after reboot
+  wait_for:
```
if previous module itself reboots and waits, then why we need this additional one?

kishen-v replied:
This was mostly an additional check on SSHd  in particular, I realise it's mostly redundant because the script would anyway use the same. 
Thanks!

mkumatag commented on install-k8s.yml:
```diff
@@ -3,6 +3,7 @@
     - masters
     - workers
   roles:
+#    - update-pkgs
```
can we have this default?

kishen-v replied:
It sometimes takes a bit of time to download the packages due to the available network speed. Hope this is okay as we introduce the step as a default one in playbooks as it may impact the time taken for the jobs to finish.

mkumatag replied:
but this improves the stability and compliance of the cluster :)

mkumatag commented on roles/update-pkgs/tasks/main.yml:
```diff
@@ -0,0 +1,27 @@
+- name: Update packages and kernel to latest available versions
+  package:
```
what will happen when this run on the RHEL os? where is the rhn registration happening?

kishen-v replied:
Moved in block that deals with the registration of the OS to the update task..

kishen-v commented:
Tested the changes on a nodes that do not require updates and a node that needed updates.
```PLAY [Install Runtime and Kubernetes] *************************************************************************************************************************************
....

TASK [update-pkgs : Update packages and kernel to latest available versions] **********************************************************************************************
ok: [10.20.183.90]
ok: [10.20.183.177]
changed: [10.20.182.197]

TASK [update-pkgs : Check if reboot required] *****************************************************************************************************************************
changed: [10.20.183.90]
fatal: [10.20.182.197]: FAILED! => {"changed": true, "cmd": "needs-restarting -r", "delta": "0:00:00.458608", "end": "2025-03-06 11:25:29.065666", "msg": "non-zero return code", "rc": 1, "start": "2025-03-06 11:25:28.607058", "stderr": "", "stderr_lines": [], "stdout": "Core libraries or services have been updated since boot-up:\n  * dbus\n  * glibc\n  * kernel\n  * systemd\n\nReboot is required to fully utilize these updates.\nMore information: https://access.redhat.com/solutions/27943", "stdout_lines": ["Core libraries or services have been updated since boot-up:", "  * dbus", "  * glibc", "  * kernel", "  * systemd", "", "Reboot is required to fully utilize these updates.", "More information: https://access.redhat.com/solutions/27943"]}
...ignoring
changed: [10.20.183.177]

TASK [update-pkgs : Reboot if necessary] **********************************************************************************************************************************
skipping: [10.20.183.177]
skipping: [10.20.183.90]
changed: [10.20.182.197]

TASK [runtime : Loading Modules Boot Time] ********************************************************************************************************************************
ok: [10.20.183.177]
ok: [10.20.183.90]
ok: [10.20.182.197]
```
Works as intended without the `wait_for` or the `ping -c ip-address` commands.

mkumatag reviewed (approved):
/lgtm
